torch/_inductor/cudagraph_trees.py

```diff
@@ -1803,6 +1838,56 @@ def _run(self, new_inputs: List[Tensor], function_id: FunctionID):
             self.try_end_curr_warmup(function_id)
 
+        # [Note: Optimistic Mutation Check]
+        # To determine whether applying cudagraph, we need to check input mutations,
+        # falling into four categories: a) no mutation, b) mutation on parameters/buffers,
+        # c) mutation on cudagraph recorded tensors, d) mutation on non-cudagraph recorded tensors.
+        # We can apply cudagraph for type a,b,c but cannot for type d. This input mutation types
+        # depends on function, current_node, and inputs.
+        #
+        # Since `check_for_mutation` is slow, there is a trade-off on making type c or d faster.
+        # - To make type d) faster, we want to `check_for_mutation` and call eager function early.
+        #   However, this adds unnecessary overhead to type a, b, c due to the extra check.
+        # - To make type c) faster, we want to skip `check_for_mutation` at the beginning and
+        #   only `check_for_mutation` before `record_function` for a new function. This removes
+        #   the overhead of `check_for_mutation` for type a, b, c. However, this adds extra overhead
+        #   to type d due to `check_invariants` for all children nodes.
+        #
+        # Instead, we design optimistic mutation check. The assumption is that, given a function
+        # and a node, the input mutation types usually remain the same across inputs. So, if we have
+        # ever detect a function on a node with type d, we will never detect it as type c. The detailed
+        # design is:
+        # - [Slow Path] On the first invocation of a function on a node, we run `check_for_mutation`
+        #   once and cache the input mutation type as `non_cudagraph_managed_mutation[node_id][func_id]`.
+        # - [Fast Path] On the subsequent invocations of a function on a node, we skip
+        #   `check_for_mutation`. For `non_cudagraph_managed_mutation[node_id][func_id]` as true, we
+        #   directly call eager function. Otherwise, we `check_variants` and call cudagraph function.
+        # - [Slow Path] Before `record_function`, we run `check_for_mutation` again.
+        #
+        # Q1: Would there be overhead for type a,b,c,d?
+        # A: No. We only check input mutation types for the first invocation of a function on a node.
+        #
+        # Q2: If a function happens to be type c during the first invocation on a node, could we detect
+        #     it as type d in the future?
+        # A:  Yes. This is done by `check_invariants` and guarantees the correctness.
+        #
+        # Q3: If a function happens to type d during the first invocation on a node, could it still
+        #     be recognized as type c in the future?
+        # A: No. But this should happen rarely according to our assumption. In the rare case that
+        #    it happens, there would not be any correctness issues and the performance is the same
+        #    as the eager (or inductor optimized) function.
+        #
```

Comment:
The explainer is helpful, thank you ! But one nit: maybe let's remove the detailed design comment ? I want to try to keep the code somewhat concise, and IMO the detailed design part is sort of redundant with the code itself and the above explanation.